src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs

```diff
@@ -86,31 +86,63 @@ internal InfiniteScrollList(Lazy<JoinableTaskFactory> joinableTaskFactory)
 
             BindingOperations.EnableCollectionSynchronization(Items, _list.ItemsLock);
 
-            ItemsView = new CollectionViewSource() { Source = Items };
-            ListCollectionView itemsView = ItemsView.View as ListCollectionView;
+            ItemsView = new CollectionViewSource() { Source = Items }.View;
+            ICollectionViewLiveShaping itemsView = (ICollectionViewLiveShaping)ItemsView;
             itemsView.IsLiveFiltering = true;
             itemsView.IsLiveGrouping = true;
             itemsView.LiveFilteringProperties.Add(nameof(PackageItemViewModel.IsPackageVulnerable));
             itemsView.LiveGroupingProperties.Add(nameof(PackageItemViewModel.PackageLevel));
-            itemsView.Filter = item =>
+            ItemsView.Filter = item =>
             {
-                if (item is not null && item.Equals(_loadingVulnerabilitiesStatusIndicator))
-                {
-                    return FilterByVulnerabilities && !(_loadingVulnerabilitiesStatusIndicator.Status == LoadingStatus.NoItemsFound && VulnerablePackagesCount > 0);
-                }
-                PackageItemViewModel vitem = item as PackageItemViewModel;
-                return IsItemInFilter(vitem);
+                return IsItemAcecpted(item);
             };
 
-            DataContext = ItemsListCollection;
+            DataContext = itemsView;
             CheckBoxesEnabled = false;
 
             _loadingStatusIndicator.PropertyChanged += LoadingStatusIndicator_PropertyChanged;
         }
 
+
+        private bool IsItemAcecpted(object item)
+        {
+            if (item is LoadingStatusIndicator)
+            {
+                return IsLoadingIndicatorAccepted(item);
+            }
+
+            if (item is PackageItemViewModel vm)
+            {
+                return IsVulnerablePackageAccepted(vm);
+            }
+
+            return false;
+        }
+
+        private bool IsLoadingIndicatorAccepted(object item)
+        {
+            if (item is not null && item.Equals(_loadingStatusIndicator))
+            {
+                return !_filterByVulnerabilities;
+            }
+
+            if (item is not null && item.Equals(_loadingVulnerabilitiesStatusIndicator))
+            {
+                return _filterByVulnerabilities && !(_loadingVulnerabilitiesStatusIndicator.Status == LoadingStatus.NoItemsFound && VulnerablePackagesCount > 0);
+            }
+
+            return false;
+        }
+
+        private bool IsVulnerablePackageAccepted(PackageItemViewModel item)
+        {
+            return !_filterByVulnerabilities
+                || (_filterByVulnerabilities && item is PackageItemViewModel vm && vm.IsPackageVulnerable);
```

Comment:
You're casting item to PackageItemViewModel but the method signature already tells us this is a PackageItemViewModel. Personally, I'd opt for casting within this method, but if you're going to have the outer method casting before we enter this IsVulnerablePackageAccepted method, we don't need to cast again in here.